src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php

```diff
@@ -75,25 +76,36 @@ final class ProductFormDataProvider implements FormDataProviderInterface
      */
     private $contextLangId;
 
+    /**
+     * @var int
+     */
+    private $shopId;
+
     /**
      * @param CommandBusInterface $queryBus
      * @param bool $defaultProductActivation
      * @param int $mostUsedTaxRulesGroupId
-     * @param int $defaultCategoryId
      * @param int $contextLangId
+     * @param int $defaultCategoryId
+     * @param int|null $contextShopId
      */
     public function __construct(
         CommandBusInterface $queryBus,
         bool $defaultProductActivation,
         int $mostUsedTaxRulesGroupId,
         int $defaultCategoryId,
-        int $contextLangId
+        int $contextLangId,
+        int $defaultShopId,
+        ?int $contextShopId
     ) {
         $this->queryBus = $queryBus;
         $this->defaultProductActivation = $defaultProductActivation;
         $this->mostUsedTaxRulesGroupId = $mostUsedTaxRulesGroupId;
-        $this->defaultCategoryId = $defaultCategoryId;
         $this->contextLangId = $contextLangId;
+        $this->defaultCategoryId = $defaultCategoryId;
+
+        // Usually context shop ID is null when multistore feature is disabled, in this case we use the default shop as fallback
+        $this->shopId = $contextShopId ?: $defaultShopId;
```

Comment:
IMHO context should always return the default shopId even if multi-store is disabled. I like to think that multi-store is always there, but we may or may not use it. WDYT?

Reply:
Well actually it's because of how the shop context service behaves:
https://github.com/PrestaShop/PrestaShop/blob/824691139ee1ee220fc1a2f3b9d7b4fed5cd33f0/src/Adapter/Shop/Context.php#L60

And more precisely how the legacy Shop also behaves:

https://github.com/PrestaShop/PrestaShop/blob/7fb9828bec07e3de4bef4f0f949ceec2b765cc4c/classes/shop/Shop.php#L1025-L1031

In the service definition I rely on this method:
https://github.com/PrestaShop/PrestaShop/blob/7bd9e1810a7a870d62786927ecb3bbd82dcb0281/src/PrestaShopBundle/Resources/config/services/core/form/form_data_provider.yml#L152-L161

So what happens when you are in context ALL_SHOP is that `Shop::$context_id_shop` and `Shop::$context_id_shop_group` are both null (which makes sense in a way, we are not in single shop mode so we can't know what sop is selected in the context). So in this case I need to decide on which shop to fallback on, and my main was actually that in tests environment the value was null as well and caused bugs so the CI was broken

But while I answer this I actually think I have a better idea here, I introduced a special shop constraint for product and can use `ProductShopConstraint::defaultProductShop()` it means it uses the shop that has been assign as default to the product (in DB it's `product.id_shop_default` basically, from what I understood, this field is filled with either:
- the shop in context when the product is created/imported in a single shop context
- the default shop otherwise (multishop disabled or all shop context)

Anyway this product's default shop is "dynamic" as it can be different from one product to another, so I think it would be the best candidate for fallback in this case.

Reply:
> Anyway this product's default shop is "dynamic" as it can be different from one product to another, so I think it would be the best candidate for fallback in this case.

Yes, good idea 👍

Reply:
I think I'll do the same in `ProductFormDataHandler` (and so in the command builders as well) It's better to rely on a shop constraint indicating the product's default rather than on a shopId which would fall back to the global default shop.

Because nothing guarantees that the global default shop is the same as the product's default shop.

Reply:
yes, totally! it should be possible to delete the product from a single context, so if one has:

Default shop: A

Product in the Shop A
Product in the Shop B

and deletes Product in the Shop B

and we have a problem

Reply:
Mmm, I think there is more a problem when you delete product from Shop A (which is the default one)

Indeed, but I'll handle this use case in a later PR, when I will introduce the command `RemoveProductFromShop` (naming not decided yet), in which case we will handle the automatic redefinition of the product's default shop, to make sure the data stays consistent This command will be handled in the multishop header new component that helps with the association with other shops

As for the listing, I don't know yet how we will handle the removal. One possibility would that we rely on this `RemoveProductFromShop` command, and when it detects that no shops are associated to the product then it removes it for good (either soft delete or hard delete depending on the product's associations)

Reply:
Done! As we said the now the provider uses the product's own default shop as a fallback when no context shop is selected Same for update commands (along all the command builder chain).

I just kept the global default shop ID for the CreateProductCommand, since the product can't have its own default shop when being created, the global default shop seems like a good candidate for fallback.